Source/WTF/wtf/MallocSpan.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <utility>
+#include <wtf/FastMalloc.h>
+#include <wtf/Noncopyable.h>
+#include <wtf/StdLibExtras.h>
+
+// MallocSpan is a smart pointer class that wraps a std::span and calls fastFree in its destructor.
+
+namespace WTF {
+
+template<typename T, typename Malloc = FastMalloc> class MallocSpan {
+    WTF_MAKE_NONCOPYABLE(MallocSpan);
+public:
+    MallocSpan() = default;
+
+    MallocSpan(MallocSpan&& other)
+        : m_span(other.leakSpan())
+    {
+    }
+
+    ~MallocSpan()
+    {
+        Malloc::free(m_span.data());
+    }
+
+    MallocSpan& operator=(MallocSpan&& other)
+    {
+        MallocSpan ptr = WTFMove(other);
+        swap(ptr);
+        return *this;
+    }
+
+    void swap(MallocSpan& other)
+    {
+        std::swap(m_span, other.m_span);
+    }
+
+    std::span<const T> span() const { return spanConstCast<const T>(m_span); }
+    std::span<T> mutableSpan() { return m_span; }
+    std::span<T> leakSpan() { return std::exchange(m_span, std::span<T>()); }
```

Comment:
Hm, I considered suggesting that this should be called `leakData` because the std::span itself is notably not leaked. But then the name does not parallel the other two functions `span` and `mutableSpan`, and surely those should not be renamed. So I suppose it's probably best as-is.

Reply:
How about `releaseSpan()`? As it is “released” from being managed by the `MallocSpan` that contains it. There are already other uses of the “release” verb so I suppose it might work.

Reply:
I think is the right term here as it is leaking. We don't usually release for this kind of thing.
There is a lot of precedent for "leak": `MallocPtr::leakPtr()`, `RefPtr::leakRef()`, `Ref::leakRef()`.

We have the concern of `release` in WebKit but it usually doesn't leak memory, e.g. `RefPtr::releaseNonNull()`, which returns a Ref.

Reply:
Can we add WARN_UNUSED_RETURN like we do for other leak functions?